### ISSUE TITLE
Reproduce personachat baselines with seq2seq v2

### ISSUE THREAD

**Issue by Adamell** [facebookresearch/ParlAI#2293]
Hello,

I am struggling to reproduce results of seq2seq model that are listed in personachat paper for 'No Persona' variant. I was able to reproduce them easily with legacy seq2seq v0 agent on personachat:none dataset, but with seq2seq (v2) agent I couldn't.

When training legacy agent seems to converge much faster and get expected resuts on validation set (ppl ~38), while the newest seq2seq agent stops at validtion ppl ~50-60. 

After few fails with seq2seq v2 agent I tried playing with parameters, augmenting training set with reversed dialogs (from other speaker perspective), but i couldn't get expected results (i always stuck at ppl 45-70).

My legacy agent training script is inspired by [the one published by convai2 team](https://github.com/facebookresearch/ParlAI/blob/master/projects/convai2/baselines/seq2seq/train.py):

```
from parlai.scripts.train_model import setup_args, TrainLoop

if __name__ == '__main__':
    parser = setup_args()
    parser.set_defaults(
        task='personachat:none',
        model='legacy:seq2seq:0',
        model_file='run_1',
        dict_lower=True,
        dict_include_valid=False,
        dict_maxexs=-1,
        datatype='train',
        batchsize=64,
        hiddensize=1024,
        embeddingsize=256,
        attention='general',
        numlayers=2,
        rnn_class='lstm',
        learningrate=3,
        dropout=0.1,
        gradient_clip=0.1,
        lookuptable='enc_dec',
        optimizer='sgd',
        embedding_type='glove',
        momentum=0.9,
        bidirectional=False,
        context_length=-1,
        validation_every_n_secs=90,
        validation_metric='ppl',
        validation_metric_mode='min',
        validation_patience=12,
        log_every_n_secs=10,
        dict_tokenizer='split',
    )
    TrainLoop(parser).train()
```

I train seq2seq v2 with following script:

```
from parlai.scripts.train_model import setup_args, TrainLoop

if __name__ == '__main__':
    parser = setup_args()
    parser.set_defaults(
        task='personachat:none',
        model='seq2seq',
        dict_lower=True,
        dict_include_valid=False,
        dict_maxexs=-1,
        datatype='train',
        batchsize=64,
        hiddensize=1024,
        embeddingsize=256,
        attention='general',
        numlayers=2,
        rnn_class='lstm',
        learningrate=3,
        dropout=0.1,
        gradient_clip=0.1,
        lookuptable='enc_dec',
        optimizer='sgd',
        embedding_type='glove',
        momentum=0.9,
        bidirectional=False,
        context_length=-1,
        validation_every_n_secs=90,
        rank_candidates=False,
        validation_metric='ppl',
        validation_metric_mode='min',
        validation_patience=12,
        log_every_n_secs=10,
        dict_tokenizer='split',
    )
    TrainLoop(parser).train()
```

I am training on 1 Tesla K80, processing about 2000 examples / 10 s log. 

I will be grateful for any help / suggestions / explanation you can provide.

**Comment by stephenroller** [comment]
Can you training upload the training logs to some gist or pastebin?

**Comment by Adamell** [comment]
@stephenroller 
Here are training logs for parameters given above:
[Legacy seq2seq v0 training](https://pastebin.com/pg7qXHUm)
[Seq2seq v2 training](https://pastebin.com/JmVhKUVm)

Probably, worth noting is fact that we were working on a private fork of parlai ([last commit](https://github.com/facebookresearch/ParlAI/commit/36d0e980)) with no modifications to core code in google colab env.

To be sure that this behavior also appears in current version of parlai we just runned seq2seq training on up to date parlai cloned master. Unfortunately we are still not able to reproduce results. Here are logs:
[Seq2seq v2 training](https://pastebin.com/3HfyiGJq)

I noticed that adding strong dropout helps a bit (with dropout 0.5 i got ppl 45). Our 'playing with parameters' also included reducing/increasing lr, increasing evaluation frequency, playing with optimizer. Nothing helped.

**Comment by stephenroller** [comment]
So for the record, we don't try to maintain compatibility between the old versions, that's why the old versions exist.

But from my memory, this stands out as a major difference that could easily be making the difference:

Old version:
https://github.com/facebookresearch/ParlAI/blob/078d09d8d3da0b1df969d73256d32c35b9132ece/parlai/agents/legacy_agents/seq2seq/seq2seq_v0.py#L483-L485

New version:
https://github.com/facebookresearch/ParlAI/blob/078d09d8d3da0b1df969d73256d32c35b9132ece/parlai/core/torch_agent.py#L1143-L1146


That is, the old one uses a gaussian matrix for projecting the embeddings from 300 -> 256 dimensional. The new one also divides by 256 to keep things in the right scale. My intuition is actually that the new way is better (hence why we added it), but I could see that failing. To test, comment out the `self.proj_rp /= target_dim` line and see if you still observe the difference.

**Comment by stephenroller** [comment]
We’re you able to make progress on this?

**Comment by Adamell** [comment]
Thank you for your concern! The change you suggested hadn't worked on its own, but put me in right direction, and after further analysis of differences between v0 and v2, I noticed 2 more changes that were applied to output layer of the model:

1. Addition of biases in o2e and e2s layers
https://github.com/facebookresearch/ParlAI/blob/master/parlai/agents/seq2seq/modules.py#L503
https://github.com/facebookresearch/ParlAI/blob/master/parlai/agents/seq2seq/modules.py#L528

2. Weights initialization as sampling from normal distribution (mean 0, std 1)
https://github.com/facebookresearch/ParlAI/blob/master/parlai/agents/seq2seq/modules.py#L509-L511

It looks like after reverting those 3 changes, I managed to get close enough to my goal with convai2 default args. I guess that maybe more extensive search of parameters for seq2seq v2 could work as well. Thank you for help!
